### PR TITLE
Centralize event states in rust-selectors

### PR DESCRIPTION
**Pull request by bholley**
This still needs a rev bump on rust-selectors once https://github.com/servo/rust-selectors/pull/55 gets merged.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8162)

### CODE REVIEW

**Comment by bholley** [comment]
Added rust-selectors rev bump.

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -80,26 +83,16 @@ use style::properties::{PropertyDeclaration, PropertyDeclarationBlock, parse_sty
...
-        const IN_HOVER_STATE = 0x04,
-        #[doc = "Content is enabled (and can be disabled)."]
-        const IN_ENABLED_STATE = 0x08,
-        #[doc = "Content is disabled."]
-        const IN_DISABLED_STATE = 0x10,
-    }
-}
+// NB: We can't [derive(..)] this because it's defined in rust-selectors, which
+// doesn't know about JS<T>.
+impl JSTraceable for EventState { fn trace(&self, _trc: *mut JSTracer) {} }
```
Comment (Manishearth) [inline components/script/dom/element.rs]: `no_jsmanaged_fields!` is what we should use here (grep the source for it)

File: components/util/mem.rs [hunk 2]
```diff
@@ -324,6 +325,12 @@ impl HeapSizeOf for () {
     }
 }
 
+impl HeapSizeOf for EventState {
```
Comment (Manishearth) [inline components/util/mem.rs]: Stick this in known_heap_size at the bottom

**Comment by Manishearth** [comment]
Reviewed 15 of 15 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlinputelement.rs, line 245 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8162#-K1Gic5ZZJ8IroKurc61)** ([raw file](https://github.com/servo/servo/blob/eabcb38ad608f7024bfc9fe22e5391c3885e6426/components/script/dom/htmlinputelement.rs#L245)):</sup>
There's some extra logic in update_checked_state, let's move the set_state there


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8162)

**Comment by bholley** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/dom/htmlinputelement.rs, line 245 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8162#-K1Gic5ZZJ8IroKurc61)** ([raw file](https://github.com/servo/servo/blob/eabcb38ad608f7024bfc9fe22e5391c3885e6426/components/script/dom/htmlinputelement.rs#L245)):</sup>
Whoops, good catch!


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8162)

**Comment by bholley** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/dom/htmlinputelement.rs, line 245 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8162#-K1Gic5ZZJ8IroKurc61)** ([raw file](https://github.com/servo/servo/blob/eabcb38ad608f7024bfc9fe22e5391c3885e6426/components/script/dom/htmlinputelement.rs#L245)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8162)

**Comment by SimonSapin** [comment]
r-

This change is unjustified, seems unnecessary to me (is there a benefit I don’t see?), and the corresponding rust-selectors change (https://github.com/servo/rust-selectors/pull/55) forces users to either keep all state in a single `u16` of bit flags like Servo does, or do mork work than necessary gathering all state when only one bit is used.

Please explain motivation for this change and the rust-selectors one.

**Comment by bholley** [comment]
@SimonSapin:
See https://github.com/servo/rust-selectors/pull/55#issuecomment-150398896 , and the email I just sent.

**Comment by SimonSapin** [comment]
Can we discuss in public, where people other than me can find it later?

**Comment by bholley** [comment]
@SimonSapin I need to run, and there's not an easy way to convert an email thread to a github issue thread. But feel free to post/summarize, and I can respond in a few hours when I get back.

**Comment by bholley** [comment]
A bit more context here now that I have a few minutes (sorry for the delay):

My high-level goal here is to fix https://github.com/servo/servo/issues/6942 by implementing restyle hints in Servo. This logic needs to live at the intersection of layout, style, and rust-selectors, and therefore will understandably involve changes to rust-selectors.

Restyle hints work by batching up a series of EventState changes for a given Element and then asking the style system for an upper-bound of the categories of elements (self, subtree, siblings, etc) that may need restyling. In order to ask this question, rust-selectors and layout need a common bitfield type by which to represent a set of event states. Given the dependency tree, this type needs to live in rust-selectors, hence this change.

Once we have this common representation, it seems like an obvious simplification to remove the per-state methods from the trait definition and pass this information with a single method. This removes a lot of boilerplate code (duplicated across tree.rs, element.rs, and wrapper.rs), and makes it possible to access the entire set of state values with a single virtual call. More importantly, it eliminates the chicken-and-egg dependency problem that causes Servo to break when rust-selectors adds support for a new EventState. This bit me today with https://github.com/servo/servo/pull/8142 , and I wanted to prevent it from biting other people.

The one downside is the one mentioned in [1], whereby other theoretical consumers without a centralized EventState representation will pay a performance cost when aggregating their state. However, optimizing performance for multiple consumer designs is pretty hard. Any consumer that wants maximal performance is going to want to centralize EventState to take advantage of restyle hints (and subsequent optimizations beyond it), so I'm not convinced that this should block what we do here.

[1] https://github.com/servo/servo/pull/8162#issuecomment-150399038

**Comment by bholley** [comment]
@Ms2ger explained to me that our use of generics allows us to bypass virtual calls on the trait impls (which is super neat!). I think the rest of the points are still valid though.

**Comment by SimonSapin** [comment]
https://github.com/SimonSapin/kuchiki is not theoretical. It doesn’t have many users yet (as far as I know?), but there was demand for it repetitively.

I’m not buying the preformance thing. As you mentioned @Ms2ger explained, the compiler generates separate code for generic functions/methods for each concrete type arguments. Moving the `bitfield.contains()` call to rust-selectors might not even affect the generated code (assuming sufficiently advanced inlining). Dynamic dispatch (and type ereasure / heterogenous collections) can be opted into by using [trait objects](https://doc.rust-lang.org/book/trait-objects.html) instead of generics.

Repetitive code probably wouldn’t be hard to generate with `macro_rules!`.

Please explain motivation in pull request messages. Restyle hints does make https://github.com/servo/rust-selectors/pull/55 seem less gratuitous than it first did in isolation.

Still, it’s not clear to me: why does rust-selectors need to know about this bitfield?

**Comment by bholley** [comment]
> Please explain motivation in pull request messages.

Sorry. The motivation is explained in commit messages and comments of other dependent commits, but I was trying to get this piece landed separately, since it requires me to update Servo's rust-selectors dependency, and I wanted to isolate any issues that might be caused by that step.

The larger context here is that @pcwalton and I are investigating what it would take to plug Servo's style system into Gecko. In order to familiarize myself with Servo's style system (and Rust in general), I'm implementing restyle hints.

> Still, it’s not clear to me: why does rust-selectors need to know about this bitfield?

When testing restyle hints, we need to match against selectors while ignoring the subset of pseudo-selectors corresponding to the EventStates that changed. Suppose a selector |S| of the form |a b:invalid:focus:hover c| exists in a stylesheet, and we change the focus _and_ hover state of an element |foo|. The restyle hint algorithm would then want to match against the effective selector |T| of the form |a b:invalid| in order to determine whether the event changes in |foo| might affect the set of elements in the document matching |S|.

However, |T| may not exist in any stylesheet, and it cannot be computed before match time because it depends on the set of EventStates that changed. So we need to pass an optional bitfield to selector matching instructing rust-selectors which pseudo-classes to ignore. This is why rust-selectors needs EventState (or something isomorphic), and my primary motivation for moving it.

Another side-benefit is that it consolidates the two existing enumerates of states (EventState and SimpleSelector) into the same library, making it easier to change both of them together (see my complaint about :target). Landing coordinated mutually-dependent changes to multiple libraries is a barrier to contribution that I'd like to remove.

> Repetitive code probably wouldn’t be hard to generate with macro_rules!

Potentially yes, but we still need to enumerate those states in all of the places I mentioned, making us susceptible to cross-library breakage.

All that being said, coalescing the API was an attempt to simplify the API and improve the cross-library ergonomic footguns that I was encountering. We could revert that change if you like without scuttling restyle hints, though I do think that doing so would make our lives harder on the whole.

**Comment by SimonSapin** [comment]
Please see my proposed changes in https://github.com/servo/rust-selectors/pull/56 and discuss them. Let’s block this in the mean time.

**Comment by bholley** [comment]
Basing this rev on @SimonSapin's PR, but will need a rev and possibly a rebase once that actually lands.

r? @SimonSapin in the mean time on everything else.

**Comment by bholley** [comment]
@SimonSapin this should now be pointing to a proper revision of rust-selectors, and ready for review.

**Comment by bholley** [comment]
review-carry pls

**Comment by Manishearth** [comment]
Probably should start using p=1 for PRs which have a tendency to bitrot often. (`@bors-servo p=1` works)

**Comment by bholley** [comment]
@Manishearth r?

**Comment by mbrubeck** [comment]
@bors-servo retry

#8157

**Pull request merged**